text/0000-cargo-binary-dependencies.md

```diff
@@ -0,0 +1,146 @@
+
+Cargo also allows depending on `cdylib` or `staticlib` artifacts. For example, you can embed a dynamic library in your binary:
+
+```rust
+// main.rs
+const MY_PRELOAD_LIB: &[u8] = include_bytes!(env!("CARGO_CDYLIB_FILE_MYPRELOAD"));
+```
+
+Note that cargo cannot help you ensure these artifacts are available at runtime for an installed version of a binary; cargo can only supply these artifacts at build time. Runtime requirements for installed crates are out of scope for this change.
+
+If you need to depend on multiple variants of a crate, such as both the binary and library of a crate, you can supply an array of strings for `type`: `type = ["bin", "lib"]`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+There are four `type`s available:
+1. `"lib"`, the default
+2. `"bin"`, a crate building one or more binaries
+3. `"cdylib"`, a C-compatible dynamic library
+4. `"staticlib"`, a C-compatible static library
+
+`"lib"` corresponds to all crates that can be depended on currently,
+including `lib`, `rlib`, and `proc-macro` libraries.
+See [linkage](https://doc.rust-lang.org/reference/linkage.html) for more information.
+
+Artifact dependencies can appear in any of the three sections of dependencies (or in target-specific versions of these sections):
+- `[build-dependencies]`
+- `[dependencies]`
+- `[dev-dependencies]`
+
+By default, `build-dependencies` are built for the host, while  `dependencies` and `dev-dependencies` are built for the target. You can specify the `target` attribute to build for a specific target, such as `target = "wasm32-wasi"`; a literal `target = "target"` will build for the target even if specifing a build dependency. (If the target is not available, this will result in an error at build time, just as if building the specified crate with a `--target` option for an unavailable target.)
+
+Cargo provides the following environment variables to the crate being built:
+
+- `CARGO_<TYPE>_DIR_<CRATE>`, where `<TYPE>` is the `type` of the artifact (uppercased) and `<CRATE>` is the package of the crate being depended on. (As with other Cargo environment variables, crate names are converted to uppercase, with dashes replaced by underscores.) This is the directory containing all the artifacts from the crate.
+- `CARGO_<TYPE>_FILE_<CRATE>_<ARTIFACT>`, where `<TYPE>` is the `type` of the artifact, `<CRATE>` is the package of the crate being depended on, and `<ARTIFACT>` is the name of the artifact. (Note that `<ARTIFACT>` is *not* modified in any way from the `name` specified in the crate supplying the artifact, or the crate name if not specified; for instance, it may be in lowercase, or contain dashes.) This is the full path to the artifact.
+    - For the crate types `cdylib` and `staticlib` that can (currently) only build one artifact, cargo additionally supplies this variable with the `_<ARTIFACT>` suffix omitted.
```

Comment:
If we ever wanted to add an ability for packages to output multiple `cdylib` or `staticlib` artifacts (which the "(currently)" suggests we might), this functionality would actively prevent us from doing so.

Reply:
@nagisa We talked about that in advance, and this does not block future such extensions. If we add a mechanism to build multiple cdylib or staticlib artifacts, those additional artifacts will have names, and the above environment variable pattern will work for those names.

That said, I've modified the RFC to change this case in a way that addresses this in advance. Rather than supplying a variable with the artifact suffix omitted only for cdylib and staticlib, cargo will supply a variable with the artifact suffix omitted whenever the artifact suffix is the same as the crate name.